components/autofill/src/db/store.rs

```diff
@@ -2,178 +2,60 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use crate::db::models::address::{Address, UpdatableAddressFields};
-use crate::db::models::credit_card::{CreditCard, UpdatableCreditCardFields};
+use crate::db::models::address::{Address, InternalAddress, UpdatableAddressFields};
+use crate::db::models::credit_card::{CreditCard, InternalCreditCard, UpdatableCreditCardFields};
 use crate::db::{addresses, credit_cards, AutofillDb};
 use crate::error::*;
+use crate::sync::engine::ConfigSyncEngine;
 use rusqlite::{
     types::{FromSql, ToSql},
     Connection,
 };
 use sql_support::{self, ConnExt};
-use std::cell::RefCell;
 use std::path::Path;
-use std::sync::{Arc, Mutex, Weak};
-use sync15_traits::SyncEngine;
+use std::sync::{Arc, Mutex};
 use sync_guid::Guid;
 
 // Our "sync manager" will use whatever is stashed here.
 lazy_static::lazy_static! {
     // Mutex: just taken long enough to update the inner stuff - needed
-    //        to wrap the RefCell as they aren't `Sync`
-    // RefCell: So we can replace what it holds. Normally you'd use `get_ref()`
-    //          on the mutex and avoid the RefCell entirely, but that requires
-    //          the mutex to be declared as `mut` which is apparently
-    //          impossible in a `lazy_static`
-    // [Arc/Weak]<StoreImpl>: What the sync manager actually needs.
-    pub static ref STORE_FOR_MANAGER: Mutex<RefCell<Weak<StoreImpl>>> = Mutex::new(RefCell::new(Weak::new()));
+    //        to wrap the Option<Store> as they aren't `Sync`
+    static ref STORE_FOR_MANAGER: Mutex<Option<Store>> = Mutex::new(None);
```

Comment:
Actually, this change bothers me a little - it's no longer a weak reference. IIUC, once `register_with_sync_manager()` is called the instance will become immortal?

Reply:
One interesting thing I recently learned about `Weak` references in Rust is that, while they don't prevent the `Drop` impl from being called, they *do* prevent freeing of the underlying memory allocation (because the weakref count itself is stored in the allocation). So the allocation here will be immortal regardless, but if we have interesting `Drop` behaviours that we'd expect to be run, then this change would prevent those from running.

Reply:
huh - that *is* surprising. However, I'm thinking less of the allocation itself but the chain of structs. Eg, the IIUC, `AutofillDb` instance would be immortal (with an open connection to the db file) as I didn't see anything setting the `Option<Store>` to `None`, so an `Arc<>` to the DB remains.

Reply:
Yeah, this was the one thing I wasn't sure about.  It's not going to be dropped, unless you register a different store.  I guess I was thinking that if a consumer did create a store, call `register_with_sync_manager()`, and dropped their reference, maybe it makes sense for it to stay registered.

Still, it's definitely a behavior change and could be risky, so I can just change it back to a Weak.